src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -0,0 +1,557 @@
+                        .Construct(F.SpecialType(SpecialType.System_Boolean)),
+                    GeneratedNames.MakeAsyndIteratorPromiseOfValueOrEndFieldName(), isPublic: true);
+
+                // Add a field: bool promiseIsActive
+                _promiseIsActiveField = F.StateMachineField(
+                    F.SpecialType(SpecialType.System_Boolean),
+                    GeneratedNames.MakeAsyndIteratorPromiseIsActiveFieldName(), isPublic: true);
+
+                // Add a field: T current
+                _currentField = F.StateMachineField(method.IteratorElementType, GeneratedNames.MakeIteratorCurrentFieldName());
+            }
+
+            /// <summary>
+            /// Generates the body of the replacement method, which initializes the state machine. Unlike regular async methods, we won't start it.
+            /// </summary>
+            protected override BoundStatement GenerateStateMachineCreation(LocalSymbol stateMachineVariable, NamedTypeSymbol frameType)
+            {
+                // PROTOTYPE(async-streams): TODO review this (what is this error case at the start?)
+
+                // If the async method's result type is a type parameter of the method, then the AsyncTaskMethodBuilder<T>
+                // needs to use the method's type parameters inside the rewritten method body. All other methods generated
+                // during async rewriting are members of the synthesized state machine struct, and use the type parameters
+                // structs type parameters.
+                AsyncMethodBuilderMemberCollection methodScopeAsyncMethodBuilderMemberCollection;
+                if (!AsyncMethodBuilderMemberCollection.TryCreate(F, method, null, out methodScopeAsyncMethodBuilderMemberCollection))
+                {
+                    return new BoundBadStatement(F.Syntax, ImmutableArray<BoundNode>.Empty, hasErrors: true);
+                }
+
+                var bodyBuilder = ArrayBuilder<BoundStatement>.GetInstance();
+                LocalSymbol builderVariable = F.SynthesizedLocal(methodScopeAsyncMethodBuilderMemberCollection.BuilderType, null);
+
+                // local.$builder = System.Runtime.CompilerServices.AsyncTaskMethodBuilder<typeArgs>.Create();
+                bodyBuilder.Add(
+                    F.Assignment(
+                        F.Field(F.Local(stateMachineVariable), _builderField.AsMember(frameType)),
+                        F.StaticCall(
+                            null,
+                            methodScopeAsyncMethodBuilderMemberCollection.CreateBuilder)));
+
+                // local.$stateField = NotStartedStateMachine;
+                bodyBuilder.Add(
+                    F.Assignment(
+                        F.Field(F.Local(stateMachineVariable), stateField.AsMember(frameType)),
+                        F.Literal(StateMachineStates.NotStartedStateMachine)));
+
+                // builder = local.$stateField.builder;
```

Comment:
>builder [](start = 19, length = 7)

Are we using this local?

Reply:
Good catch. Thanks
In an async method, we make this local and then use it to start the builder (background process). That's not needed for async-iterator method.

---
In reply to: [214171546](https://github.com/dotnet/roslyn/pull/28218#discussion_r214171546) [](ancestors = 214171546)